network/ClientNetworking.cpp

```diff
@@ -127,13 +128,23 @@ ClientNetworking::ClientNetworking() :
     m_io_service(),
     m_socket(m_io_service),
     m_incoming_messages(m_mutex),
-    m_connected(false),
-    m_cancel_retries(false)
+    m_rx_connected(false),
+    m_tx_connected(false)
 {}
 
-bool ClientNetworking::Connected() const {
+bool ClientNetworking::IsConnected() const {
     boost::mutex::scoped_lock lock(m_mutex);
-    return m_connected;
+    return m_rx_connected && m_tx_connected;
+}
+
+bool ClientNetworking::IsRxConnected() const {
+    boost::mutex::scoped_lock lock(m_mutex);
```

Comment:
Here is an example.  Assume two threads, A and B, are trying to read a single bit.  

The bit is stored in RAM on a capacitor C, storing either a +1 or -1 unit of charge.  It is read by discharging it into an amplifier to determine if it is a +1 or a -1 and then recharging the capacitor to the correct level.

Assume that the capacitor starts with C = +1 and both thread A and thread B are reading it with no mutex.

1. To start with the capacitor is charged with one unit of charge, C = +1
2. Thread A starts reading and discharges the capacitor into its amplifier.  Thread A reads the correct value +1. Now C = 0
3. Thread B starts reading, but there is no charge on the capacitor, so its amplifier doesn't receive enough charge and will amplify a random value.  Thread B reads a random value R and the capacitor is still C=0
4. Thread A and thread B are now racing to see who writes back to the capacitor first.
a)  Thread A writes back the correct value and C = +1.  This is your best case scenario, only one bad read happened, but that random value will propagate through the code.
b)  Thread B writes back the random value R.  Now every subsequent read will see the random value.
c) Thread A and B write back opposite values and the cap is floating at some random value, C ~= 0.  Now the next read will see a different random value.


The mutex protects from this problem.

You can also see why the problem is insidious.